### ISSUE TITLE
API for 3rd party support

### ISSUE THREAD

**Issue by wanseob** [zkopru-network/zkopru#156]
**Is your feature request related to a problem? Please describe.**
Zkopru node should provide an API for the 3rd party support.

**Describe the solution you'd like**

Create a new package `client` and provides the following functions

```typescript
[WIP]
interface ZkopruClient {
    setProvider(provider: L2Provider): void
    getProvider(): L2Provider
    isConnected(): Promise<boolean>
    node: ZkopruNode
    layer1: Layer1
    layer2: Layer2
    wallet: ZkWallet
}

/**
 * JsonRpcProvider: Communicate with Zkopru node using json-rpc
 * ContextProvider: Zkopru node runs in the same process.
 */
type L2Provider = JsonRpcProvider | ContextProvider

interface ZkopruNode {
    isRunning(): Promise<boolean>
    isSynced(): Promise<boolean>
    start(): Promise<boolean>
    stop(): Promise<boolean>
}

interface Layer1 {
    getVKs(): Promise<{ [txSig: string]: VerifyingKey }>
    getConfig(): Promise<Config>
    sendTx<T>(tx: TransactionObject<T>, account: Account, to?: string, option?: Tx): Promise<TransactionReceipt  undefined>
}

interface Layer2 {
    latestBlock(): Promise<Bytes32>
    getBlock(hash: Bytes32): Promise<Block | null>
    getProposal(hash: Bytes32): Promise<Proposal>
    getDeposits(...massDeposits: MassDeposit[]): Promise<Deposit[]>
    getPendingMassDeposits(): Promise<PendingMassDeposits>
    getOldestUnprocessedBlock(): Promise<Unprocessed | undefined>
    isUncleBlock(parentBlock: Bytes32, proposalNum: number): Promise<boolean>
    applyBootstrap(block: Block, bootstrapData: BootstrapData)
    getGrovePatch(block: Block): Promise<GrovePatch>
    getTokenRegistry(): Promise<TokenRegistry>
}

interface ZkWallet {
    getAccounts(): Promise<ZkAccount[]>
    setDefaultAccount(account: number | ZkAccount): void
    createAccount(idx: number): Promise<void>
    getSpendableAmount(account?: ZkAccount): Promise<NoteSum>
    getLockedAmount(account?: ZkAccount): Promise<NoteSum>
    getSpendables(account?: ZkAccount): Promise<Utxo[]>
    getWithdrawables(account?: ZkAccount, status?: WithdrawalStatus): Promise<Withdrawal[]>
    getUtxos(account?: ZkAccount, status?: UtxoStatus): Promise<Utxo[]>
    fetchLayer1Assets(account?: ZkAccount): Promise<Balance>
    listLayer1Nfts(erc721Addr: string, balance: number, account?: ZkAccount): Promise<string[]>
    depositEther(eth: F, fee: F, to?: ZkAddress): Promise<boolean>
    depositERC20(eth: F, addr: string, amount: F, fee: F, to?: ZkAddress): Promise<boolean>
    depositERC721(eth: F, addr: string, nft: F, fee: F, to?: ZkAddress): Promise<boolean>
    withdraw(withdrawal: Withdrawal): Promise<boolean>
    instantWithdrawal(prePayer: Address, withdrawal: Withdrawal): Promise<boolean>
    fetchPrice(): Promise<string>
    sendTx(tx: RawTx, from?: ZkAccount, encryptTo?: ZkAddress): Promise<boolean>
    getTxBuilder(account?: ZkAccount): TxBuilder
}
```

**Additional context**

**Comment by JChanceHud** [comment]
@wanseob Should this be a http/ws interface for accessing these functions? It would be nice to expose the client package as a wrapper for connecting to this for things like block explorers.

I can take this task if you're looking for someone to do this.

**Comment by wanseob** [comment]
Yeah what I was thinking is using `ZkopruClient` just like the "web3" & making the connection between the client & zkopru node via HTTP, Websocket or ipc. Because we're using typescript, implementing json-rpc vis HTTP will be the first option for us. How do you think? And it'll be so happy for me if you take this task :smile:

**Comment by JChanceHud** [comment]
Yes, I'm thinking of defining a rpc protocol and then building the different transports on top (ws/http/ipc), just like geth.

For the `ZkopruNode` part there needs to be a fast way to sync data from the `L2Provider`. If there are lots of transactions in the history this data could be quite large. Some way of compressing and sending historical data as one large file might work well. Once a `ZKWallet` library exists we should write a script to send lots of transactions for a few days so we could see how the size of the chain scales.

I think it makes sense to split the Zkopru client like this:

```ts
interface ZKClient {
    setProvider(provider: L2Provider): void
    getProvider(): L2Provider
    isConnected(): Promise<boolean>
    layer1: Layer1
    layer2: Layer2
}
```

```ts
interface ZKWallet extends ZKClient {
    node: ZkopruNode
    getAccounts(): Promise<ZkAccount[]>
    setDefaultAccount(account: number | ZkAccount): void
    createAccount(idx: number): Promise<void>
    getSpendableAmount(account?: ZkAccount): Promise<NoteSum>
    getLockedAmount(account: ZkAccount): Promise<NoteSum>
    getSpendables(account?: ZkAccount): Promise<Utxo[]>
    getWithdrawables(account?: ZkAccount, status?: WithdrawalStatus): Promise<Withdrawal[]>
    getUtxos(account?: ZkAccount, status?: UtxoStatus): Promise<Utxo[]>
    fetchLayer1Assets(account?: ZkAccount): Promise<Balance>
    listLayer1Nfts(erc721Addr: string, balance: number, account?: ZkAccount): Promise<string[]>
    depositEther(eth: F, fee: F, to?: ZkAddress): Promise<boolean>
    depositERC20(eth: F, addr: string, amount: F, fee: F, to?: ZkAddress): Promise<boolean>
    depositERC721(eth: F, addr: string, nft: F, fee: F, to?: ZkAddress): Promise<boolean>
    withdraw(withdrawal: Withdrawal): Promise<boolean>
    instantWithdrawal(prePayer: Address, withdrawal: Withdrawal): Promise<boolean>
    fetchPrice(): Promise<string>
    sendTx(tx: RawTx, from?: ZkAccount, encryptTo?: ZkAddress): Promise<boolean>
    getTxBuilder(account?: ZkAccount): TxBuilder
}
```

This way a web client could embed only `ZKClient` in a webpage for a block explorer while keeping the bundle light.

**Comment by wanseob** [comment]
> For the ZkopruNode part there needs to be a fast way to sync data from the L2Provider. If there are lots of transactions in the history this data could be quite large. Some way of compressing and sending historical data as one large file might work well. Once a ZKWallet library exists we should write a script to send lots of transactions for a few days so we could see how the size of the chain scales.

Sounds very cool. I was considering the light node design with some "bootstrapping" but your idea seems more general to me.
> I think it makes sense to split the Zkopru client like this:

Agree with you, it will help people to build block explorers!

**Comment by JChanceHud** [comment]
This is the RPC spec I'm thinking of implementing. In addition to these functions I think it would be nice if the node proxied eth requests to the l1 node. e.g. If the method is `eth_call` it's passed to the geth node if proxy mode is enabled. This would allow a zkopru url to be used as a web3 provider.

I think this is a good starting point for the API. It should give us what we need to expand later. This spec is based on a more general spec i wrote [here](https://github.com/JChanceHud/rollup-api/blob/master/SPEC.md).

The work for this issue is in #164 

## l1_address

Returns the address of the current L2 contract on the L1 chain.

- [ ] implemented

### Parameters

None

### Returns

`DATA` - 20 bytes contract address

## l1_getVKs

Return the verifying keys from the L1 contract.

- [x] implemented

### Parameters

None

### Returns

Public key data

## l2_blockNumber

Returns the highest L2 block number broadcasted and confirmed on the L1 chain.

- [x] implemented

### Parameters

None

### Returns

`NUMBER` - Integer of the current block number

## l2_getBlockByNumber

Returns block data by number.

- [x] implemented

### Parameters

1. `NUMBER|TAG` - integer of a block number, or the string `earliest` or `latest`.

```
params: [
  "0x1C8"
]
```

### Returns

Block data

## l2_getBlockByHash

Returns block data by hash.

- [x] implemented

### Parameters
1. `DATA` - bytes hash of block in hex

### Returns

`Object` - A block object or `null` if no block is found

## l2_getEthBalance

Returns the Ether balance for a given address in the L2 network. @wanseob lmk if this isn't possible.

- [ ] implemented

### Parameters
1. `DATA` - 20 bytes - Address to check
2. `NUMBER|TAG` - block number of the string `earliest` or `latest` - Balance at which block

### Returns

`NUMBER` - Balance in wei

## l2_getERC20Balance

Returns the ERC20 token balance for a given address in the L2 network.

- [ ] implemented

### Parameters

1. `DATA` - 20 bytes - Address to check balance
2. `DATA` - 20 bytes - L1 address of the ERC20 token contract
3. `NUMBER|TAG` - block number of the string `earliest` or `latest` - Balance at which block

### Returns

`NUMBER` - Token balance with decimals

### Example

```js
// Request
{
  "method": "l2_getERC20Balance",
  "params": [
    "0x407d73d8a49eeb85d32cf465507dd71d507100c1",
    "0x6B175474E89094C44Da98b954EedeAC495271d0F", // DAI contract
    "latest"
  ],
  "jsonrpc": "1.0",
  "id": "e91e6a69-b7f2-42ce-9e66-be562bbc5dec"
}

// Response
{
  "result": "0xD02AB486CEDC0000", // 15 DAI
  "id": "e91e6a69-b7f2-42ce-9e66-be562bbc5dec",
  "jsonrpc": "1.0"
}
```

## l2_getTransactionByHash

Returns data for an L2 transaction.

- [ ] implemented

### Parameters
1. `DATA` - bytes - The transaction hash

### Returns

`Object` - Transaction object

## l2_getRegisteredTokens

Returns a list of token addresses that are registered in the system. This could return ERC20 and ERC721 tokens with flags determining which is which.

- [ ] implemented

### Parameters

None

### Returns

`Array` - Array of contract addresses

**Comment by wanseob** [comment]
> This is the RPC spec I'm thinking of implementing. In addition to these functions I think it would be nice if the node proxied eth requests to the l1 node. e.g. If the method is eth_call it's passed to the geth node if proxy mode is enabled. This would allow a zkopru url to be used as a web3 provider.

I love this idea a lot

I think we need to divide API into two parts

1. General API (for example, APIs for block explorer)
2. Personal APIs: This should contain L2-getBalance and functions described in the `ZkWallet` interface above.

**Comment by JChanceHud** [comment]
I agree, but I wonder if a personal RPC API is necessary? Wouldn't a wallet be running their own validator and thus be able to call functions directly?

I think building a library to run a wallet is important and should have the `ZkWallet` interface I just don't see why it should be accessible remotely.

**Comment by wanseob** [comment]
IMO they'll be accessed via localhost, then we can run a node & build an app that runs on another process separately via the rpc call. I thought it's similar to web3.personal api. Does this make sense?

**Comment by JChanceHud** [comment]
Yes that makes sense. The `ZkWallet` interface can just be connected to the RPC and optionally enabled.

I'll probably first build the general API, then `ZkWallet`, then the RPC for `ZkWallet`.

**Comment by wanseob** [comment]
@JChanceHud 's PR #164 is resolving this issue but I'll keep this open since we still need to implement the `ZkWallet` interface.
